tool/planet/dns.go

```diff
@@ -110,16 +107,48 @@ func updateEnvDNSAddresses(client *kubernetes.Clientset, role agent.Role) error
 	return trace.Wrap(writeEnvDNSAddresses([]string{svcWorker.Spec.ClusterIP, svcMaster.Spec.ClusterIP}, true))
 }
 
-// createService creates the kubernetes DNS service if it doesn't already exist.
-// The service object is managed by gravity, but we create a placeholder here, so that we can read the IP address
-// of the service, and configure kubelet with the correct DNS addresses before starting
-func createService(name string) error {
+func createDNSServices(ctx context.Context, serviceCIDR net.IPNet) error {
 	client, err := cmd.GetKubeClientFromPath(constants.SchedulerConfigPath)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	ipalloc := ipallocator.NewAllocatorCIDRRange(&serviceCIDR)
+	services := client.CoreV1().Services(metav1.NamespaceSystem)
+	for _, name := range []string{"kube-dns", "kube-dns-worker"} {
+		if err := createDNSService(ctx, name, services, ipalloc); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
 
-	service := &v1.Service{
+// createDNSService creates the kubernetes DNS service if it doesn't already exist.
+// The service object is managed by gravity, but we create a placeholder here, so that we can read the IP address
+// of the service, and configure kubelet with the correct DNS addresses before starting
+func createDNSService(ctx context.Context, name string, services corev1.ServiceInterface, ipalloc *ipallocator.Range) error {
+	ip, err := ipalloc.AllocateNext()
+	if err != nil && err != ipallocator.ErrFull {
+		return &backoff.PermanentError{Err: err}
+	}
+	return utils.RetryWithInterval(ctx, newUnlimitedExponentialBackoff(5*time.Second), func() error {
+		_, err = services.Create(newDNSService(name, ip.String()))
+		if err == nil || errors.IsAlreadyExists(err) {
```

Comment:
So if we're ignoring "already exists" error, what's taking care of recreating the DNS services from the new IP range? Is it done by Gravity?

Reply:
Yes, gravity manages the DNS services during the operation. The agent could in theory ensure that the DNS service with an IP address from the service cidr existed and forced it if not but there's also a race if an agent on a node where container configuration is yet to be updated restarts, it might invalidate the operation state so I chose to keep the modifications on one side.